diff --git a/cmake/modules/extensions.cmake b/cmake/modules/extensions.cmake
index 1eef042e88c..7b11238c585 100644
--- a/cmake/modules/extensions.cmake
+++ b/cmake/modules/extensions.cmake
@@ -434,6 +434,11 @@ macro(zephyr_library)
   zephyr_library_named(${lib_name})
 endmacro()
 
+macro(zephyr_app_library)
+  zephyr_library_get_current_dir_lib_name(${ZEPHYR_BASE} lib_name)
+  zephyr_app_library_named(${lib_name})
+endmacro()
+
 # Determines what the current directory's lib name would be according to the
 # provided base and writes it to the argument "lib_name"
 macro(zephyr_library_get_current_dir_lib_name base lib_name)
@@ -469,6 +474,18 @@ macro(zephyr_library_named name)
   target_link_libraries(${name} PUBLIC zephyr_interface)
 endmacro()
 
+macro(zephyr_app_library_named name)
+  # This is a macro because we need add_library() to be executed
+  # within the scope of the caller.
+  set(ZEPHYR_CURRENT_LIBRARY ${name})
+  add_library(${name} STATIC "")
+
+  zephyr_append_cmake_library(${name})
+  set_property(GLOBAL APPEND PROPERTY ZEPHYR_APP_LIBS ${name})
+
+  target_link_libraries(${name} PUBLIC zephyr_interface)
+endmacro()
+
 # Provides amend functionality to a Zephyr library for out-of-tree usage.
 #
 # Usage:
diff --git a/cmake/modules/kernel.cmake b/cmake/modules/kernel.cmake
index af9c85d80e3..886b08cae6f 100644
--- a/cmake/modules/kernel.cmake
+++ b/cmake/modules/kernel.cmake
@@ -34,6 +34,12 @@ define_property(GLOBAL PROPERTY ZEPHYR_LIBS
 zephyr_library() appends libs to this list.")
 set_property(GLOBAL PROPERTY ZEPHYR_LIBS "")
 
+define_property(GLOBAL PROPERTY ZEPHYR_APP_LIBS
+    BRIEF_DOCS "Global list of all Zephyr CMake libs that should be linked with application (app target)"
+    FULL_DOCS  "Global list of all Zephyr CMake libs that should be linked with application (app target).
+zephyr_app_library() appends libs to this list.")
+set_property(GLOBAL PROPERTY ZEPHYR_APP_LIBS "")
+
 define_property(GLOBAL PROPERTY ZEPHYR_INTERFACE_LIBS
     BRIEF_DOCS "Global list of all Zephyr interface libs that should be linked in."
     FULL_DOCS  "Global list of all Zephyr interface libs that should be linked in.
